extensions/sha256/circuit/src/sha256_chip/mod.rs

```diff
@@ -0,0 +1,203 @@
+//! Sha256 hasher. Handles full sha256 hashing with padding.
+//! variable length inputs read from VM memory.
+use std::{
+    array,
+    cmp::{max, min},
+    sync::Arc,
+};
+
+use openvm_circuit::{
+    arch::{ExecutionBridge, ExecutionBus, ExecutionError, ExecutionState, InstructionExecutor},
+    system::{
+        memory::{MemoryControllerRef, MemoryReadRecord, MemoryWriteRecord},
+        program::ProgramBus,
+    },
+};
+use openvm_circuit_primitives::{bitwise_op_lookup::BitwiseOperationLookupChip, encoder::Encoder};
+use openvm_instructions::{
+    instruction::Instruction,
+    program::DEFAULT_PC_STEP,
+    riscv::{RV32_CELL_BITS, RV32_REGISTER_NUM_LIMBS},
+    UsizeOpcode,
+};
+use openvm_rv32im_circuit::adapters::read_rv32_register;
+use openvm_sha256_air::{Sha256Air, SHA256_BLOCK_BITS};
+use openvm_sha256_transpiler::Rv32Sha256Opcode;
+use openvm_stark_backend::p3_field::PrimeField32;
+use sha2::{Digest, Sha256};
+
+mod air;
+mod columns;
+mod trace;
+
+pub use air::*;
+pub use columns::*;
+#[cfg(test)]
+mod tests;
+
+// ==== Constants for register/memory adapter ====
+/// Register reads to get dst, src, len
+const SHA256_REGISTER_READS: usize = 3;
+/// Number of cells to read in a single memory access
+const SHA256_READ_SIZE: usize = 16;
+/// Number of cells to write in a single memory access
+const SHA256_WRITE_SIZE: usize = 32;
```

Comment:
see previous comment about write size considerations to make it more compatible with potential input into another hash